File: ios/Cove/SettingsScreen/NewPinView.swift
//
//  NewPinView.swift
//  Cove
//
//  Created by Praveen Perera on 12/12/24.
//

import SwiftUI

private enum PinState {
    case new, confirm(String)
}

struct NewPinView: View {
    /// args
    var onComplete: (String) -> Void
    var backAction: () -> Void

    /// private
    @State private var pinState: PinState = .new

    var body: some View {
        Group {
            switch pinState {
            case .new:
                NumberPadPinView(
                    title: "Enter New PIN",
                    isPinCorrect: { _ in true },
                    showPin: false,
                    backAction: backAction,
                    onUnlock: { enteredPin in
                        withAnimation {
                            pinState = .confirm(enteredPin)
                        }
                    }
                )
            case let .confirm(pinToConfirm):
                NumberPadPinView(
                    title: "Confirm New PIN",
                    isPinCorrect: { $0 == pinToConfirm },
                    showPin: false,
                    backAction: backAction,
                    onUnlock: onComplete
                )
            }
        }
    }
}

#Preview {
    NewPinView(onComplete: { _ in }, backAction: {})
}

File: android/app/src/main/java/org/bitcoinppl/cove/settings/NewPinView.kt
package org.bitcoinppl.cove.settings

import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import org.bitcoinppl.cove.views.NumberPadPinView

private sealed class NewPinState {
    data object New : NewPinState()

    data class Confirm(
        val pinToConfirm: String,
    ) : NewPinState()
}

@Composable
fun NewPinView(
    onComplete: (String) -> Unit,
    backAction: () -> Unit,
) {
    var pinState: NewPinState by remember { mutableStateOf(NewPinState.New) }

    when (val state = pinState) {
        is NewPinState.New -> {
            NumberPadPinView(
                title = "Enter New PIN",
                isPinCorrect = { true },
                showPin = true,
                backAction = backAction,
                onUnlock = { enteredPin ->
                    pinState = NewPinState.Confirm(enteredPin)
                },
            )
        }

        is NewPinState.Confirm -> {
            NumberPadPinView(
                title = "Confirm New PIN",
                isPinCorrect = { it == state.pinToConfirm },
                showPin = false,
                backAction = backAction,
                onUnlock = onComplete,
            )
        }
    }
}
